### PR TITLE
Login redesign

### PR DESCRIPTION
**Pull request by thornbill**
**Changes**
The long awaited redesign for the login workflow.
* Moves login workflow to MVVM architecture
* Uses a single activity for startup and login workflow

To Do:
* [x] Finalize data structure for users and servers
* [x] Error handling and input validation
* [x] Prevent duplicate server listings
* [x] Remove old login code

Follow Up Tasks:
* [ ] Saving servers and previously logged in users (current implementation only saves most recent login)
* [ ] Finish ui/ux of server listing screen (currently relies on existing `GridButton` implementation)
* [ ] Finish ui/ux of login screen
* [ ] PIN support
* [ ] Managing saved servers and users (option to remove/forget)
* [ ] Use kotlinx.serialization instead of gson
* [ ] Quick connect support

**Issues**

### CODE REVIEW

File: app/src/main/java/org/jellyfin/androidtv/data/source/JsonFileSource.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/source/JsonFileSource.kt:20]: FYI; The next apiclient version doesn't provide a serializer anymore, and if it did it would not use GSON but kotlinx.serialization. I'd recommend switching to kotlinx.serialization in this PR for all new JSON stuff.

Optionally; it also supports [CBOR](http://cbor.io) which might have slightly better performance.

**Comment by thornbill** [comment]
List of follow up items added to #612

**Comment by thornbill** [comment]
Build is failing due to the removed translation that I didn't remove in other locales. I'm not sure if I should wait until we're ready to merge to remove it or just rebase before we merge to avoid conflicts with weblate.

**Comment by nielsvanvelzen** [comment]
I'd say add back the removed translations and remove them in a followup pull request removing the translations from all locales that we will merge immediately.

**Comment by nielsvanvelzen** [comment]
Testing notes:

- [ ] The login screen when no server is added should show something like "no servers here yet" or maybe even show a "welcome" screen for new users. (https://f.ndat.nl/1604672131.png)
- [ ] The "enter server address" screen should submit the url when pressing the OK/ENTER button on a remote to avoid dpad navigation to the "OK" button
- [ ] After adding my own server the app returned to the login screen but nothing happened and my server was not added. I later found out this was due too an untrusted certificate. These kinds of errors should be handled [1]
- [ ] The user sign in UI inputs should be slightly bigger and use a serif font instead of monospace
- [ ] Pressing OK/ENTER should signin instead of focussing on the OK button
- [ ] After signing in the server selection screen shows up and then opens the main interface, this looks ugly. It should go directly to that screen instead.
- [ ] After restarting the app my user (that I just signed in) is gone and I'm not automatically authenticated
- [ ] In app-settings the authentication options should be updated
- [ ] The logout button in the app-home should skip the "loading" screen
- [ ] When starting the app the "loading" screen sometimes shows up for a second after the server-selection screen was shown
- [ ] How do I remove a server?
- [ ] How do I re-order my servers?
- [ ] How do I rename a server?
- [ ] I can't add the demo servers (stable and unstable) at the same time because they have the same server id


[1] server connect failure not handled:
```
2020-10-04 15:18:46.327 32015-32015/org.jellyfin.androidtv.debug E/jellyfin-apiclient: VolleyError com.android.volley.NoConnectionError: javax.net.ssl.SSLHandshakeException: Chain validation failed
    com.android.volley.NoConnectionError: javax.net.ssl.SSLHandshakeException: Chain validation failed
        at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:181)
        at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:131)
        at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:111)
        at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:90)
     Caused by: javax.net.ssl.SSLHandshakeException: Chain validation failed
        at com.android.org.conscrypt.ConscryptFileDescriptorSocket.startHandshake(ConscryptFileDescriptorSocket.java:231)
        at com.android.okhttp.internal.io.RealConnection.connectTls(RealConnection.java:196)
        at com.android.okhttp.internal.io.RealConnection.connectSocket(RealConnection.java:153)
        at com.android.okhttp.internal.io.RealConnection.connect(RealConnection.java:116)
        at com.android.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:186)
        at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:128)
        at com.android.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:97)
        at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:289)
        at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:232)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:465)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponse(HttpURLConnectionImpl.java:411)
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponseCode(HttpURLConnectionImpl.java:542)
        at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getResponseCode(DelegatingHttpsURLConnection.java:106)
        at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:30)
        at com.android.volley.toolbox.HurlStack.executeRequest(HurlStack.java:96)
        at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:123)
        at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:131) 
        at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:111) 
        at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:90) 
     Caused by: java.security.cert.CertificateException: Chain validation failed
        at com.android.org.conscrypt.TrustManagerImpl.verifyChain(TrustManagerImpl.java:719)
        at com.android.org.conscrypt.TrustManagerImpl.checkTrustedRecursive(TrustManagerImpl.java:549)
        at com.android.org.conscrypt.TrustManagerImpl.checkTrustedRecursive(TrustManagerImpl.java:570)
        at com.android.org.conscrypt.TrustManagerImpl.checkTrustedRecursive(TrustManagerImpl.java:615)
        at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:505)
        at com.android.org.conscrypt.TrustManagerImpl.checkTrusted(TrustManagerImpl.java:424)
        at com.android.org.conscrypt.TrustManagerImpl.getTrustedChainForServer(TrustManagerImpl.java:352)
        at android.security.net.config.NetworkSecurityTrustManager.checkServerTrusted(NetworkSecurityTrustManager.java:94)
        at android.security.net.config.RootTrustManager.checkServerTrusted(RootTrustManager.java:89)
        at com.android.org.conscrypt.Platform.checkServerTrusted(Platform.java:220)
        at com.android.org.conscrypt.ConscryptFileDescriptorSocket.verifyCertificateChain(ConscryptFileDescriptorSocket.java:407)
        at com.android.org.conscrypt.NativeCrypto.SSL_do_handshake(Native Method)
        at com.android.org.conscrypt.NativeSsl.doHandshake(NativeSsl.java:386)
        at com.android.org.conscrypt.ConscryptFileDescriptorSocket.startHandshake(ConscryptFileDescriptorSocket.java:226)
        at com.android.okhttp.internal.io.RealConnection.connectTls(RealConnection.java:196) 
        at com.android.okhttp.internal.io.RealConnection.connectSocket(RealConnection.java:153) 
        at com.android.okhttp.internal.io.RealConnection.connect(RealConnection.java:116) 
        at com.android.okhttp.internal.http.StreamAllocation.findConnection(StreamAllocation.java:186) 
        at com.android.okhttp.internal.http.StreamAllocation.findHealthyConnection(StreamAllocation.java:128) 
        at com.android.okhttp.internal.http.StreamAllocation.newStream(StreamAllocation.java:97) 
        at com.android.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:289) 
        at com.android.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:232) 
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.execute(HttpURLConnectionImpl.java:465) 
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponse(HttpURLConnectionImpl.java:411) 
        at com.android.okhttp.internal.huc.HttpURLConnectionImpl.getResponseCode(HttpURLConnectionImpl.java:542) 
        at com.android.okhttp.internal.huc.DelegatingHttpsURLConnection.getResponseCode(DelegatingHttpsURLConnection.java:106) 
        at com.android.okhttp.internal.huc.HttpsURLConnectionImpl.getResponseCode(HttpsURLConnectionImpl.java:30) 
        at com.android.volley.toolbox.HurlStack.executeRequest(HurlStack.java:96) 
        at com.android.volley.toolbox.BasicNetwork.performRequest(BasicNetwork.java:123) 
        at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:131) 
        at com.android.volley.NetworkDispatcher.processRequest(NetworkDispatcher.java:111) 
        at com.android.volley.NetworkDispatcher.run(NetworkDispatcher.java:90) 
     Caused by: java.security.cert.CertPathValidatorException: timestamp check failed
        at sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:135)
        at sun.security.provider.certpath.PKIXCertPathValidator.validate(PKIXCertPathValidator.java:222)
        at sun.security.provider.certpath.PKIXCertPathValidator.validate(PKIXCertPathValidator.java:140)
2020-10-04 15:18:46.327 32015-32015/org.jellyfin.androidtv.debug E/jellyfin-apiclient:     at sun.security.provider.certpath.PKIXCertPathValidator.engineValidate(PKIXCertPathValidator.java:79)
        at java.security.cert.CertPathValidator.validate(CertPathValidator.java:301)
        at com.android.org.conscrypt.TrustManagerImpl.verifyChain(TrustManagerImpl.java:715)
        	... 31 more
     Caused by: java.security.cert.CertificateNotYetValidException: Certificate not valid until Fri Oct 30 15:33:03 GMT+01:00 2020 (compared to Sun Oct 04 15:18:46 GMT+02:00 2020)
        at com.android.org.conscrypt.OpenSSLX509Certificate.checkValidity(OpenSSLX509Certificate.java:265)
        at sun.security.provider.certpath.BasicChecker.verifyTimestamp(BasicChecker.java:194)
        at sun.security.provider.certpath.BasicChecker.check(BasicChecker.java:144)
        at sun.security.provider.certpath.PKIXMasterCertPathValidator.validate(PKIXMasterCertPathValidator.java:125)
        	... 36 more
```

File: app/src/main/java/org/jellyfin/androidtv/ui/startup/SplashFragment.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/startup/SplashFragment.kt:14]: You can pass the layout-id to fragments and use the default `onCreateView` function like so:

```suggestion
class SplashFragment : Fragment(R.layout.fragment_startup)
```

Comment (thornbill) [reply app/src/main/java/org/jellyfin/androidtv/ui/startup/SplashFragment.kt]: :+1: Should be fixed in any classes added here.

**Comment by thornbill** [comment]
A couple notes on the testing feedback:

* [ ] How do I re-order my servers?
    * The order is currently:
      1. Current server
      2. Saved server(s) (I was thinking these would be ordered by recently used when we add support for multiple saved servers)
      3. Discovered servers
    * I'm not sure options for explicit ordering are needed?
* [ ] How do I rename a server?
    * This is the name returned by the server. I'm not sure if we need to support renaming in the client.

**Comment by nielsvanvelzen** [comment]
> I'm not sure options for explicit ordering are needed?  

I don't think it's really necessary but might be nice to have in production-testing server setups etc. Although sorted by recently used will probably work fine.

> I'm not sure if we need to support renaming in the client.  

Once again, nice to have. If I use a server from a friend who calls it "Jellyfin" I might want to rename it to "Bill's server"

File: app/src/main/java/org/jellyfin/androidtv/ui/startup/AddServerFragment.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/startup/AddServerFragment.kt:22]: See `onCreateView` comment on the parent class

File: app/src/main/java/org/jellyfin/androidtv/ui/startup/ListServerFragment.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/startup/ListServerFragment.kt:116]: You can make these subclasses of ListServerFragment

File: app/src/main/java/org/jellyfin/androidtv/ui/startup/LoginViewModel.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/startup/LoginViewModel.kt:60]: Using something like
```kt
_serverList.apply {
    addSource(_loadingCurrentServer) {				
```

will probably make it more readable, just add all addSource calls in the .apply {} block

File: app/src/main/java/org/jellyfin/androidtv/ui/startup/StartupToolbarFragment.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/startup/StartupToolbarFragment.kt:12]: See `onViewCreated` comment on the alert fragment

File: app/src/main/java/org/jellyfin/androidtv/util/apiclient/ModelExtensions.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/util/apiclient/ModelExtensions.kt:65]: The return type doesn't need to be declared for all functions in this file as they all use expression bodies.

File: app/src/main/java/org/jellyfin/androidtv/ui/startup/StartupActivity.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/startup/StartupActivity.kt:75]: We probably wan't to use `openNextActivity()` here because the user might have opened the app with an intent but required authentication first.

File: app/src/main/java/org/jellyfin/androidtv/ui/startup/StartupActivity.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/startup/StartupActivity.kt:91]:
```suggestion
		isLoaded = true
```
# 😱

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/jellyfin/androidtv/data/repository/ServerRepository.kt
Comment (thornbill) [reply app/src/main/java/org/jellyfin/androidtv/data/repository/ServerRepository.kt:44]: Yes :smile:

File: app/src/main/java/org/jellyfin/androidtv/data/repository/ServerRepository.kt
Comment (thornbill) [reply app/src/main/java/org/jellyfin/androidtv/data/repository/ServerRepository.kt:50]: That's less than ideal... it limits our ability to handle the case where users don't have a saved server, but can discover servers.

File: app/src/main/java/org/jellyfin/androidtv/data/repository/UserRepository.kt
Comment (thornbill) [reply app/src/main/java/org/jellyfin/androidtv/data/repository/UserRepository.kt:34]: I did not and did not look into it too much since this will likely change anyway with the new apiclient implementation.

File: app/src/main/AndroidManifest.xml
Comment (thornbill) [reply app/src/main/AndroidManifest.xml:45]: It _almost_ works... except the "Add Server" button is collapsed to nothing lol

File: app/src/main/java/org/jellyfin/androidtv/data/repository/ServerRepository.kt
Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/androidtv/data/repository/ServerRepository.kt:50]: In my own network it completes in around 20 seconds, with the available servers showing up within a few seconds. But when it pings a server that doesn't exist it needs to wait for a timeout to happen.

File: app/src/main/java/org/jellyfin/androidtv/data/model/LoadingState.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/data/model/LoadingState.kt:14]:
```suggestion
		fun error(@StringRes messageRes: Int? = null) = LoadingState(Status.ERROR, msgRes)
```
For consistency

File: app/src/main/java/org/jellyfin/androidtv/ui/startup/ListServerFragment.kt
Comment (nielsvanvelzen) [inline app/src/main/java/org/jellyfin/androidtv/ui/startup/ListServerFragment.kt:70]: I think this one can be moved to `onCreateView`

**Review by nielsvanvelzen — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/jellyfin/androidtv/ui/startup/ListServerFragment.kt
Comment (thornbill) [reply app/src/main/java/org/jellyfin/androidtv/ui/startup/ListServerFragment.kt:70]: The viewmodel observer seems to need to be added in `onActivityCreated` otherwise you get some weird crashes when the activity changes.

Comment (nielsvanvelzen) [reply app/src/main/java/org/jellyfin/androidtv/ui/startup/ListServerFragment.kt:70]: I only meant the padding line

**Review by nielsvanvelzen — APPROVED** [review]

**Pull request merged**
